### PR TITLE
[ARCTIC-636][optimizer] Clear useless data files produced by failed full optimize which support hive

### LINKED ISSUE #636: [Bug]: The fullmajor optimize task commit failed
### What happened?

All optimize tasks of a table are in the "prepared" state. I checked the AMS log and found that the optimize task has been completed, but every commit fails. In addition, the optimizer reported a failover before, because the resource manager restarted.

### Affects Versions

0.3.2

### What engines are you seeing the problem on?

Optimizer

### How to reproduce

step 1: optimizer failover
step 2: start new optimize task

### Relevant log output

```shell
2022-11-18 09:39:50,633 ERROR [Metastore Scheduled Commit Worker 3] [com.netease.arctic.ams.server.optimize.OptimizeCommitWorker] [] - Optimize Committer Worker-3 com.netease.arctic.ams.server.optimize.TableOptimizeItem@7b863ac5 unexpected commit error
java.lang.Exception: unexpected commit error
        at com.netease.arctic.ams.server.optimize.BaseOptimizeCommit.commit(BaseOptimizeCommit.java:145) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.SupportHiveCommit.commit(SupportHiveCommit.java:110) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.TableOptimizeItem.commitOptimizeTasks(TableOptimizeItem.java:823) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.OptimizeCommitWorker.run(OptimizeCommitWorker.java:50) [arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_261]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_261]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_261]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_261]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_261]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_261]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_261]
Caused by: com.netease.arctic.hive.exceptions.CannotAlterHiveLocationException: can't delete hive partition: Partition(values: [2022-11-17/android], location: hdfs://eadhadoop/user/da_luna/hive_db/da_luna.db/arctic_dict_homepage_outlog/hive/day=2022-11-17/client=android/1668693508469_31637), file under partition is not deleted: hdfs://eadhadoop/user/da_luna/hive_db/da_luna.db/arctic_dict_homepage_outlog/hive/day=2022-11-17/client=android/1668693508469_31637/10-B-31637-00000-593788945-0000000001.parquet
        at com.netease.arctic.hive.op.UpdateHiveFiles.checkPartitionDelete(UpdateHiveFiles.java:238) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.hive.op.UpdateHiveFiles.lambda$getDeletePartition$5(UpdateHiveFiles.java:225) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at java.util.HashMap$Values.forEach(HashMap.java:981) ~[?:1.8.0_261]
        at com.netease.arctic.hive.op.UpdateHiveFiles.getDeletePartition(UpdateHiveFiles.java:225) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.hive.op.UpdateHiveFiles.commit(UpdateHiveFiles.java:97) ~[arctic-hive-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.BaseOptimizeCommit.majorCommit(BaseOptimizeCommit.java:314) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
        at com.netease.arctic.ams.server.optimize.BaseOptimizeCommit.commit(BaseOptimizeCommit.java:129) ~[arctic-ams-server-0.3.2-SNAPSHOT.jar:?]
```

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by hzluting**
## Why are the changes needed?
fix #636 
Failed full optimization that supports Hive produced useless data files in the partition location, which is unclear.

## Brief change log

  - *before retrying failed full optimize support Hive, try to clear useless data files*
  - *optimizer add max execute time to avoid producing useless data files, and throw timeout exception*

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java [hunk 1]
```diff
@@ -192,15 +209,37 @@ public boolean canExecute(Supplier<Integer> maxRetry) {
...
     } else if (getOptimizeStatus() == OptimizeStatus.Failed) {
+      // clear useless files produced by failed full optimize task support hive
+      String location = getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
+      if (location != null) {
+        try {
+          ArcticTable arcticTable = ServiceContainer.getOptimizeService()
+              .getTableOptimizeItem(getTableIdentifier()).getArcticTable();
+          for (FileStatus fileStatus : arcticTable.io().list(location)) {
+            String fileLocation = fileStatus.getPath().toUri().getPath();
+            if (fileLocation.contains(optimizeRuntime.getAttemptId())) {
```
Comment (zhoujinsong) [inline ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java:220]: What is the content of attempt id ?
Is it safe to find out files produced by a certain task by this operation?

File: optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java [hunk 2]
```diff
@@ -50,4 +75,15 @@ protected long getMaxTransactionId(List<DataFile> dataFiles) {
 
     return 0;
   }
+
+  protected void executeTimeout(Closeable writer) throws Exception {
```
Comment (zhoujinsong) [inline optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java:79]: What about renaming this method to `checkIfTimeout`?

**Review by zhoujinsong — COMMENTED** [review]
@hzluting  I left some comments, please take a look.

File: optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java (hunk 2 above)
Comment (hzluting) [reply optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java:79]: ok

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java (hunk 1 above)
Comment (hzluting) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java:220]: The attempt id produced by Math.abs(ThreadLocalRandom.current().nextInt())

Comment (hzluting) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java]: I will use a regular expression to match the file name

Comment (zhoujinsong) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java]: Random integers seem to collide easily, where did this attempt id generate from? executor or AMS?

You can comment the file name pattern here to help others understand your code.

Comment (hzluting) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java]: This attempt id is generated by the executor.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java [hunk 3]
```diff
@@ -87,6 +94,18 @@ public void onPending() {
     try {
       Preconditions.checkArgument(optimizeRuntime.getStatus() != OptimizeStatus.Prepared,
           "task prepared, can't on pending");
+
+      // can update max execute time on optimizing
```
Comment (wangtaohz) [inline ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java:98]: We should set the `max-execute-time` after tasks are polled from the queue

File: optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java [hunk 4]
```diff
@@ -50,4 +75,15 @@ protected long getMaxTransactionId(List<DataFile> dataFiles) {
 
     return 0;
   }
+
+  protected void checkIfTimeout(Closeable writer) throws Exception {
+    long maxExecuteTime = task.getMaxExecuteTime() != null ?
+        task.getMaxExecuteTime() : TableProperties.OPTIMIZE_EXECUTE_TIMEOUT_DEFAULT;
+    if (System.currentTimeMillis() - startTime > maxExecuteTime * factor) {
+      writer.close();
+      LOG.info("table {} execute task {} timeout, max execute time is {}",
```
Comment (wangtaohz) [inline optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java:84]: the log level `ERROR` may be better here

File: optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java [hunk 5]
```diff
@@ -50,4 +75,15 @@ protected long getMaxTransactionId(List<DataFile> dataFiles) {
...
+          table.id(), task.getTaskId(), task.getMaxExecuteTime());
+      throw new TimeoutException("optimizer execute timeout");
```
Comment (wangtaohz) [inline optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java:86]: `TimeoutException`'s message should contain more information like 'actual execute time', 'max execute time', and 'factor'.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java [hunk 6]
```diff
@@ -192,15 +210,40 @@ public boolean canExecute(Supplier<Integer> maxRetry) {
     if (getOptimizeStatus() == OptimizeStatus.Init) {
       return true;
     } else if (getOptimizeStatus() == OptimizeStatus.Failed) {
+      // clear useless files produced by failed full optimize task support hive
```
[the remark this answers is not in the document]
Comment (hzluting) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java]: ok

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java (hunk 3 above)
Comment (hzluting) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/optimize/OptimizeTaskItem.java]: ok

File: optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java (hunk 5 above)
Comment (hzluting) [reply optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java]: ok

File: optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java (hunk 4 above)
Comment (hzluting) [reply optimizer/src/main/java/com/netease/arctic/optimizer/operator/executor/BaseExecutor.java]: ok

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java [hunk 7]
```diff
@@ -374,6 +377,27 @@ private void addIntoOptimizeQueue(OptimizeTaskItem task) throws InvalidObjectExc
           throw new InvalidObjectException(
               queueName() + " not allow task of table " + task.getTableIdentifier());
         }
+        // clear useless files produced by failed full optimize task support hive
+        String location = task.getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
+        if (location != null) {
+          try {
+            ArcticTable arcticTable = ServiceContainer.getOptimizeService()
+                .getTableOptimizeItem(task.getTableIdentifier()).getArcticTable();
+            for (FileStatus fileStatus : arcticTable.io().list(location)) {
```
Comment (wangtaohz) [inline ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java:386]: If location not exists, should skip deleting files.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java [hunk 8]
```diff
@@ -374,6 +377,27 @@ private void addIntoOptimizeQueue(OptimizeTaskItem task) throws InvalidObjectExc
           throw new InvalidObjectException(
               queueName() + " not allow task of table " + task.getTableIdentifier());
         }
+        // clear useless files produced by failed full optimize task support hive
+        String location = task.getOptimizeTask().getProperties().get(OptimizeTaskProperties.CUSTOM_HIVE_SUB_DIRECTORY);
```
Comment (wangtaohz) [inline ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/impl/OptimizeQueueService.java:381]: Check task status is Failed, before deleting files

**Comment by wangtaohz** [comment]
LGTM

**Pull request merged**